cmd/thanos/receive.go

```diff
@@ -321,23 +367,145 @@ func runReceive(
 
 		s := shipper.New(logger, reg, dataDir, bkt, func() labels.Labels { return lset }, metadata.ReceiveSource)
 
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+		// Before starting, ensure any old blocks are uploaded.
+		if uploaded, err := s.Sync(context.Background()); err != nil {
+			level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+		}
+		// Before quitting, ensure all blocks are uploaded.
+		defer func() {
+			if uploaded, err := s.Sync(context.Background()); err != nil {
+				level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+			}
+		}()
 
-			return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
-				if uploaded, err := s.Sync(ctx); err != nil {
-					level.Warn(logger).Log("err", err, "uploaded", uploaded)
-				}
+		{
+			// Run the uploader in a loop.
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
+					if uploaded, err := s.Sync(ctx); err != nil {
+						level.Warn(logger).Log("err", err, "uploaded", uploaded)
+					}
 
-				return nil
+					return nil
+				})
+			}, func(error) {
+				cancel()
 			})
-		}, func(error) {
-			cancel()
-		})
+		}
+
+		{
+			// Upload on demand.
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case <-uploadC:
+						if uploaded, err := s.Sync(ctx); err != nil {
+							level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+						}
+						cancel()
+						sendOrCancel(ctx.Done(), uploadDone)
+					}
+				}
+			}, func(error) {
+				cancel()
+				close(uploadDone)
+			})
+		}
 	}
 
+	// Before quitting, ensure the WAL is flushed.
+	defer func() {
+		db, err := tsdb.Open(
```

Comment:
I don't see a `db.Close()` anywhere

Reply:
:+1:

Reply:
the `db.Close()` is in the flush func

Reply:
Why not keep it in the block where the db is opened to avoid the confusion?

Reply:
Flushing has been moved to a new type with a `Flush` method. The method docstring documents this, so I think that is plenty